### ISSUE TITLE
[redux-saga] Updating libdef for 0.14.0

### ISSUE THREAD

**Issue by kuy** [flowtype/flow-typed#569]
Hi, I've just started working on libdef for [redux-saga 0.14.0](https://github.com/yelouafi/redux-saga/releases/tag/v0.14.0).

**Comment by kuy** [comment]
Memo: Enhanced for v0.14.0 https://gist.github.com/8th713/16e99834c8461abd7a780cbad15183d4

**Comment by tai2** [comment]
I tried the above gist and noticed a possible problem.

When I use redux-saga, I often write some code like below:

```
import { takeLatest } from 'redux-saga/effects';

function* worker(action) {
    console.log(action.type);
}

export default function* rootSaga() {
    yield takeLatest('AN_ACTION', worker);
}
```

Just using takeLatest(or takeEvery) with generator functions.
This is my favorite pattern because it's easy to use.

I wanted to make that pattern typed. I struggled with flow and finally succeeded. The result is here.

```
/* @flow */
import { takeLatest } from 'redux-saga/effects';
import type { Effect } from 'redux-saga/effects';

type ExampleAction = {
    type: string;
}

function* worker(action?: ExampleAction) {
    if (!action) {
        return;
    }
    console.log(action.type);
}

export default function* rootSaga(): Generator<Effect,void,*> {
    yield takeLatest('AN_ACTION', worker);
}
```

Testing action argument is forced because `Workable0` in the libdef take an action as optional.
But I know an action should be passed in in this case, testing the argument should not be required.

I'm not sure why action is typed as optional, there may be some unavoidable reasons.
In my app, I just rewrote libdef and make action parameter mandatory.

Example Repo: https://github.com/tai2/redux-saga-take-latest-typing-example

**Comment by ironblock** [comment]
I don't know if it will help anybody here, but I came across this issue while trying to solve my own coverage gap. I've started working on an updated libdef for `v0.15.x` based on the .d.ts files from , and hopefully I can open a PR soon. If anybody has work already started, I'd be very happy to incorporate it.

**Comment by GAntoine** [comment]
@ironblock Thanks for taking the time to update it. There a little tool called [flowgen](https://github.com/joarwilk/flowgen) which can auto translate .d.ts files, with some success. Could save you some time translating what's left.

**Comment by ironblock** [comment]
@GAntoine yeah, that's a good catch. I started from the `v0.13.x` file already present, the `v0.14.x` file posted in this thread, and flowgen. Sadly, flowgen doesn't seem like it wants to play nice with the official TypeScript files. I found that a lot of output looks like this, and requires a lot of manual tweaking:
```javascript
declare export function all(effects: Effect[]): AllEffectdeclare export type GenericRaceEffectDescriptor<T>= {
    [key: string]: T
};
```

In the end, I used it as a third reference point for "what's one answer for how this might be translated?". There are also a number of places where the official `.d.ts` does something that Flow doesn't support (as far as I can tell), e.g.:
```typescript
export type SagaIterator = IterableIterator<Effect|Effect[]>;
```

This type of thing is currently getting my best approximation, and I'm looking forward to feedback and insight on the inevitable PR. The good news is that I absolutely need this for work, since it causes a huge gap in Flow coverage. We use `redux-saga` for all API interaction, and Flow doesn't seem to understand that the reference passed into `call()` is going to be invoked. Blech.

**Comment by thehogfather** [comment]
@ironblock is

> ```javascript
> export type SagaIterator = IterableIterator<Effect|Effect[]>;
>```

equivalent to this?
> ```javascript
> type _Effect = Effect | Effect[]
> export type SagaIterator = IterableIterator<_Effect>;
>```

**Comment by 102** [comment]
Whoever starts working on this issue, be aware, that current version of `redux-saga` is `0.16.0`
